feat: add multi-tenant support --story=125575006

Zhaoyikaiii opened the pull request.

Zhaoyikaiii committed:
feat: add multi-tenant support and configuration for InfluxDB integration

Zhaoyikaiii committed:
feat: add tenant-aware key generation for Redis lookups

chenjiandongx commented on pkg/unify-query/influxdb/spacetsdb_router_test.go:
```diff
@@ -15,14 +15,14 @@ import (
 	"testing"
 	"time"
 
+	miniredis "github.com/alicebob/miniredis/v2"
```
原来的 redis 方案是有什么问题吗？

Zhaoyikaiii replied:
替换为miniRedis的目的是能够在测试用例里面直接跑起来一个redis而不需要依赖本地的reidis

chenjiandongx commented on pkg/unify-query/influxdb/settings.go:
```diff
@@ -0,0 +1,18 @@
+package influxdb
+
+const (
+	SystemTenantWithSuffixConfigPath = "influxdb.system_tenant_with_suffix"
```
1）为什么这里只有 influxdb 有？
2）是不是可以换个 option 名称，比如 `启动系统租户特性` 之类的。

Zhaoyikaiii replied:
1) 按早上说的我会修改为:victoria_metrics.system_tenant.enable_suffix.原先为influxdb是因为这个包是influxdb
2) 修改为:victoria_metrics.system_tenant.enable_suffix

chenjiandongx commented on pkg/unify-query/influxdb/spacetsdb_router.go:
```diff
@@ -38,6 +39,25 @@ var (
+	user := metadata.GetUser(ctx)
+	tenantID := user.TenantID
+
+	if tenantID != "" {
```
编写代码建议还是优先判断更短路径，让代码看起来尽量层级少一些。

如比这里可以优先判断 tenantID  为空的情况。

Zhaoyikaiii replied:
已修改:[fix: add missing gettenantSuffixKey callfix: add missing](https://github.com/TencentBlueKing/bkmonitor-datalink/pull/855/commits/632b7319e6a21911a41b28620608cd61c4e23355)

chenjiandongx commented on pkg/unify-query/influxdb/spacetsdb_router.go:
```diff
@@ -429,7 +451,8 @@ func (r *SpaceTsDbRouter) GetResultTable(ctx context.Context, tableID string, ig
 // GetDataLabelRelatedRts 获取 DataLabel 详情，仅包含映射的 RT 信息
 func (r *SpaceTsDbRouter) GetDataLabelRelatedRts(ctx context.Context, dataLabel string) influxdb.ResultTableList {
-	genericRet := r.Get(ctx, influxdb.DataLabelToResultTableKey, dataLabel, true, false)
+	key := getTenantSuffixKey(ctx, dataLabel)
```
<img width="865" alt="image" src="https://github.com/user-attachments/assets/5c9afbdd-6c08-41c1-ba54-99c94b269c02" />

是 FieldRelated 不用修改？

Zhaoyikaiii replied:
已添加

Zhaoyikaiii committed:
chore: change config variable 'SysTenantWithSuffixConfigPath' config
path

- influxdb.system_tenant_With_suffix ->
  victoria_metrics.system.enable_suffix

Zhaoyikaiii committed:
fix: add missing gettenantSuffixKey callfix: add missing
gettenantSuffixKey call

chenjiandongx commented on pkg/unify-query/influxdb/settings.go:
```diff
@@ -0,0 +1,18 @@
+package influxdb
+
+const (
+	SystemTenantWithSuffixConfigPath = "victoria_metrics.system_tenant.enable_suffix"
```
但这个代码的入口还是在 influxdb pkg 下？

Zhaoyikaiii replied:
需要保留在influxdb包下面,如果移到http包.在使用的时候会有循环依赖.

shamcleren commented on pkg/unify-query/influxdb/spacetsdb_router.go:
```diff
@@ -38,6 +39,24 @@ var (
+// It appends the tenant ID from the context to the original key using a '|' separator.
+// For the "system" tenant, the suffix is only added if the SystemTenantWithSuffix feature flag is enabled.
+func getTenantSuffixKey(ctx context.Context, key string) string {
+	user := metadata.GetUser(ctx)
```
SystemTenantWithSuffix 判定应该放到这里

Zhaoyikaiii replied:
已修改

shamcleren commented on pkg/unify-query/influxdb/spacetsdb_router.go:
```diff
@@ -438,7 +460,8 @@ func (r *SpaceTsDbRouter) GetDataLabelRelatedRts(ctx context.Context, dataLabel
 
 // GetFieldRelatedRts 获取 Field 指标详情，仅包含映射的 RT 信息
 func (r *SpaceTsDbRouter) GetFieldRelatedRts(ctx context.Context, field string) influxdb.ResultTableList {
```
这个函数可以删掉了，没在用，同类的还有 channel 等等

Zhaoyikaiii replied:
已移除: [移除GetFieldRelatedRts)](https://github.com/TencentBlueKing/bkmonitor-datalink/pull/855/commits/4d0451b2a38abd4ea00daac086a33e223c5c910c )

shamcleren commented on pkg/unify-query/unify-query.yaml:
```diff
@@ -37,6 +37,7 @@ influxdb:
   max_limit: 1e8
   max_slimit: 1e5
   tolerance: 5
+  system_tenant_with_suffix: false
```
这里应该是是否开启多租户判定

Zhaoyikaiii replied:
已修改: [更新unify-query.yaml](https://github.com/TencentBlueKing/bkmonitor-datalink/pull/855/commits/8bc768b78eaf9f360ad6af1b5c4ea80acce14a7f)

Zhaoyikaiii committed:
refactor: change config path

- remove SystemtenantWithSuffix
- add function flag with control Unify-query tenantMode
- config MultiTenantMode by using config path `http.multiTenantMode`

Zhaoyikaiii committed:
refactor: remove FieldToResultTable related code

Zhaoyikaiii committed:
chore: change unify-query.yaml

- update MultiTenantMode config path

Zhaoyikaiii committed:
Merge branch 'master' into feat/multitenant-space-routing/#1010158081125575006

shamcleren committed:
Merge branch 'master' into feat/multitenant-space-routing/#1010158081125575006

Zhaoyikaiii committed:
feat: 添加测试用例以支持不完整的 tableID 处理

Zhaoyikaiii committed:
Merge branch 'master' into feat/multitenant-space-routing/#1010158081125575006

shamcleren committed:
feat: multitenant-space-routing #1010158081125575006

shamcleren committed:
feat: multitenant-space-routing #1010158081125575006

shamcleren committed:
feat: multitenant-space-routing #1010158081125575006

shamcleren committed:
feat: multitenant-space-routing #1010158081125575006

shamcleren committed:
feat: multitenant-space-routing #1010158081125575006

shamcleren committed:
Merge branch 'master' into feat/multitenant-space-routing/#1010158081125575006

shamcleren committed:
feat: multitenant-space-routing #1010158081125575006

shamcleren committed:
Merge remote-tracking branch 'Zhaoyikaiii/feat/multitenant-space-routing/#1010158081125575006' into feat/multitenant-space-routing/#1010158081125575006
